File: plans/issue-pr-bugfix.scala
//> using dep "io.github.riccardomerolla::llm4zio-runner:2.7.1"
//> using scala "3.8.3"
//> using jvm 21

/** Bug-report → fix flow for a Scala project — the ZIO-native counterpart of
  * orca's `issue-pr-bugfix.sc`.
  *
  * Given a `owner/repo#number` issue reference, the flow:
  *   1. Reads the issue from GitHub.
  *   2. Triages it (NotABug / Untestable / Testable).
  *      - NotABug / Untestable → comment the verdict and stop.
  *   3. For a Testable bug: branch, write the failing test, push, open a PR with
  *      a tentative description.
  *   4. Wait for CI to go red (fail loudly if it passes — the repro is wrong).
  *   5. Plan + implement the fix on the same branch (review per task).
  *   6. Push the fix and regenerate the PR title/body from the full diff.
  *
  * Usage: `scala-cli run issue-pr-bugfix.sc -- "acme/widgets#42"`.
  * Requires `claude` and `gh` authenticated; the repo must have CI that runs the tests.
  */

import zio.*
import java.nio.file.Path

import llm4zio.core.{CliConnectorConfig, ConnectorId}
import llm4zio.flow.*
import llm4zio.runner.Llm4zio

object Main extends ZIOAppDefault:

  // CLI for both reasoning (triage / review / PR summary) and coding — no API key.
  private val reasoning = CliConnectorConfig(ConnectorId.ClaudeCli)
  private val coder     = CliConnectorConfig(ConnectorId.ClaudeCli, flags = Map("permission-mode" -> "acceptEdits"))
  private val CiTimeout  = 30.minutes

  def run =
    getArgs.flatMap { args =>
      val workDir = Path.of(".").toAbsolutePath.normalize
      val refOpt  = args.headOption.flatMap(IssueRef.parse)

      Llm4zio.run(workDir, reasoning, coder) { ctx =>
        given FlowEvents = ctx.events
        refOpt match
          case None      => fail("usage: issue-pr-bugfix.sc -- \"owner/repo#number\"")
          case Some(ref) => bugfix(ctx, workDir, ref)
      }
    }

  private def bugfix(ctx: FlowContext, workDir: Path, ref: IssueRef)(using FlowEvents): IO[FlowError, Unit] =
    for
      issue   <- stage(s"Read issue ${ref.shortRef}")(ctx.gh.readIssue(ref))
      verdict <- stage("Triage")(Planner.triage(ctx.reasoning, issue.title, issue.body))
      _       <- verdict match
                   case Triage.NotABug(explanation) =>
                     stage("Comment: not a bug")(ctx.gh.writeIssueComment(ref, explanation))
                   case Triage.Untestable(_, steps) =>
                     stage("Comment: reproduction steps")(ctx.gh.writeIssueComment(ref, s"## Reproduction\n\n$steps"))
                   case Triage.Testable(summary, branchName, failingTestPath) =>
                     fixTestable(ctx, workDir, ref, issue, summary, branchName, failingTestPath)
    yield ()

  private def fixTestable(
    ctx: FlowContext,
    workDir: Path,
    ref: IssueRef,
    issue: Issue,
    summary: String,
    branchName: String,
    failingTestPath: String,
  )(using FlowEvents): IO[FlowError, Unit] =
    for
      _     <- stage("Branch")(ctx.git.checkoutOrCreate(branchName))
      coder <- Chat.start(ctx.coder, system = Some("You write code in the current repo."))
      _     <- stage("Write the failing test") {
                 coder
                   .ask(s"Write a failing unit test at `$failingTestPath` that reproduces: ${issue.title}\n\n${issue.body}")
                   .mapError(e => FlowError.Llm(e.toString)) *>
                   ctx.git.commitAll(s"Add failing test: $summary").unit
               }
      _       <- stage("Push branch")(ctx.git.push("origin", branchName))
      tentative <- stage("Tentative PR summary") {
                     summarisePr(
                       ctx.reasoning,
                       diff = "", // only the failing test has landed; let the model lead on the issue context
                       context = Some(s"Originating issue: ${ref.shortRef}\nTitle: ${issue.title}\n(Only a failing test has been added so far.)"),
                     )
                   }
      pr      <- stage("Open PR")(ctx.gh.createPr(tentative.title, s"${tentative.body}\n\nCloses ${ref.shortRef}."))
      status  <- stage("Wait for CI to fail")(ctx.gh.waitForBuild(pr, CiTimeout))
      _       <- ZIO.when(status == BuildOutcome.Success)(
                   fail("CI passed on the failing-test commit — the reproduction doesn't reproduce.")
                 )
      _       <- stage("Comment on PR")(ctx.gh.writePrComment(pr, s"CI is red as expected for: $summary. Implementing the fix."))
      fixPlan <- stage("Plan the fix")(
                   Planner.from(ctx.reasoning, s"Fix ${ref.shortRef} on branch $branchName so the failing test passes without regressing others.")
                 )
      planPath = workDir.resolve(s".llm4zio/fix-${ref.number}.md")
      _       <- PlanStore.save(planPath, fixPlan)
      _       <- implementTaskLoop(planPath, fixPlan) { task =>
                   for
                     _ <- coder.ask(task.description).mapError(e => FlowError.Llm(e.toString))
                     _ <- reviewAndFixLoop(Reviewers.minimal, ctx.reasoning, coder, task.title, ctx.git.diff)
                     _ <- ctx.git.commitAll(s"$branchName: ${task.title}").unit
                   yield ()
                 }
      _       <- stage("Push the fix")(ctx.git.push("origin", branchName))
      diff    <- ctx.git.diff
      finalSummary <- stage("Final PR summary")(
                        summarisePr(
                          ctx.reasoning,
                          diff,
                          context = Some(s"Issue ${ref.shortRef}: ${issue.title}\nThe branch now has the failing test AND the fix."),
                        )
                      )
      _       <- stage("Update PR")(ctx.gh.updatePr(pr, finalSummary.title, s"${finalSummary.body}\n\nCloses ${ref.shortRef}."))
      _       <- PlanStore.delete(planPath)
    yield ()

File: examples/java/IssuePrBugfix.java
//> using dep "io.github.riccardomerolla:llm4zio-java:3.19.0"
//> using scala "3.8.3"
//> using jvm 21

// Bug-report → fix flow, authored in Java — the Java-surface counterpart of issue-pr-bugfix.sc.
//
// Read + triage the issue (NotABug / Untestable are commented and the flow stops). For a Testable bug: branch, write
// the failing test, push, open a tentative PR, wait for CI to go RED (fail loudly on green — the repro is wrong), then
// plan + implement the fix (reviewed + briefed), push, and regenerate the PR title/body from the full diff.
//
// Run: scala-cli run IssuePrBugfix.java -- "owner/repo#number"
// Requires `claude` and `gh` authenticated; the repo must have CI that runs the tests.

import llm4zio.javaapi.*;
import llm4zio.flow.Triage;

public class IssuePrBugfix {
  static final long CI_TIMEOUT_SECONDS = 30 * 60;

  public static void main(String[] args) {
    Llm4zioJava.flow(args, "owner/repo#number", flow -> {
      var maybeRef = Refs.issue(flow.userPrompt());
      if (maybeRef.isEmpty()) {
        flow.fail("usage: scala-cli run IssuePrBugfix.java -- \"owner/repo#number\"");
        return;
      }
      var ref = maybeRef.get();
      var issue = flow.stage("Read issue " + ref.shortRef(), () -> flow.gh().readIssue(ref));
      var verdict = flow.stage("Triage", () -> flow.triage(issue.title(), issue.body()));

      if (verdict instanceof Triage.NotABug notABug) {
        flow.stage("Comment: not a bug", () -> flow.gh().writeIssueComment(ref, notABug.explanation()));
        return;
      }
      if (verdict instanceof Triage.Untestable untestable) {
        flow.stage("Comment: reproduction steps",
            () -> flow.gh().writeIssueComment(ref, "## Reproduction\n\n" + untestable.reproductionSteps()));
        return;
      }
      var testable = (Triage.Testable) verdict;
      var summary = testable.summary();
      var branchName = testable.branchName();
      var failingTestPath = testable.failingTestPath();

      var start = flow.git().currentBranch();
      flow.stage("Branch", () -> flow.git().checkoutOrCreate(branchName));
      var chat = flow.startChat("You write code in the current repo.");
      flow.stage("Write the failing test", () -> {
        chat.ask("Write a failing unit test at `" + failingTestPath + "` that reproduces: "
            + issue.title() + "\n\n" + issue.body());
        flow.git().commitAll("Add failing test: " + summary);
      });
      flow.stage("Push branch", () -> flow.git().push("origin", branchName));
      var tentative = flow.stage("Tentative PR summary", () -> flow.summarisePr("",
          "Originating issue: " + ref.shortRef() + "\nTitle: " + issue.title()
              + "\n(Only a failing test has been added so far.)"));
      var pr = flow.stage("Open PR",
          () -> flow.gh().createPr(tentative.title(), tentative.body() + "\n\nCloses " + ref.shortRef() + "."));
      var status = flow.stage("Wait for CI to fail", () -> flow.gh().waitForBuild(pr, CI_TIMEOUT_SECONDS));
      if (status.isSuccess()) {
        flow.fail("CI passed on the failing-test commit — the reproduction doesn't reproduce.");
      }
      flow.stage("Comment on PR",
          () -> flow.gh().writePrComment(pr, "CI is red as expected for: " + summary + ". Implementing the fix."));

      var fixPrompt = "Fix " + ref.shortRef() + " on branch " + branchName
          + " so the failing test passes without regressing others.";
      var fixPlan = flow.stage("Plan the fix (review + brief)",
          () -> flow.briefedPlan(flow.reviewedPlan(flow.planFrom(fixPrompt)), fixPrompt));
      var planPath = flow.workDir().resolve(".llm4zio/fix-" + ref.number() + ".md");
      flow.savePlan(planPath, fixPlan);
      flow.implementTaskLoop(planPath, fixPlan, task -> {
        chat.ask(fixPlan.taskPrompt(task));
        flow.reviewAndFixLoop(Reviewers.minimal(), chat, task.title(), () -> flow.git().diff());
        flow.git().commitAll(branchName + ": " + task.title());
      });
      flow.stage("Push the fix", () -> flow.git().push("origin", branchName));
      var diff = flow.git().diff();
      var finalSum = flow.stage("Final PR summary", () -> flow.summarisePr(diff,
          "Issue " + ref.shortRef() + ": " + issue.title() + "\nThe branch now has the failing test AND the fix."));
      flow.stage("Update PR",
          () -> flow.gh().updatePr(pr, finalSum.title(), finalSum.body() + "\n\nCloses " + ref.shortRef() + "."));
      flow.deletePlan(planPath);
      flow.stage("Return to " + start, () -> flow.git().checkout(start));
    });
  }
}
